sql/catalyst/src/main/java/org/apache/spark/sql/connector/write/WriteBuilder.java

```diff
@@ -35,22 +35,41 @@
 public interface WriteBuilder {
 
   /**
-   * Returns a {@link BatchWrite} to write data to batch source. By default this method throws
-   * exception, data sources must overwrite this method to provide an implementation, if the
-   * {@link Table} that creates this write returns {@link TableCapability#BATCH_WRITE} support in
-   * its {@link Table#capabilities()}.
+   * Returns a logical {@link Write} shared between batch and streaming.
+   *
+   * @since 3.2.0
    */
+  default Write build() {
+    return new Write() {
+      @Override
+      public BatchWrite toBatch() {
+        return buildForBatch();
+      }
+
+      @Override
+      public StreamingWrite toStreaming() {
+        return buildForStreaming();
+      }
+    };
+  }
+
+  /**
+   * Returns a {@link BatchWrite} to write data to batch source.
+   *
+   * @deprecated use {@link #build()} instead.
+   */
+  @Deprecated
   default BatchWrite buildForBatch() {
     throw new UnsupportedOperationException(getClass().getName() +
```

Comment:
@sunchao, I thought about the idea to default these implementations as `build().toBatch()` but it will introduce a circular dependency between these methods. There may be implementations that only implement `buildForBatch`, for example. Right now, calling `buildForStreaming` will produce an exception for them. If we default, this will result in a stackoverflow exception.

Reply:
I think you'll also need to change `WriteBuilder.build` to not override `toBatch` and `toStreaming`? such as
```scala
  default Write build() {
    return new Write() {
    };
  }
```

with this, Spark users just need to override `build` method in their DS implementations and the old APIs `buildForBatch` and `buildForStreaming` will automatically pick up the logic in the other methods.

Reply:
@sunchao, I think because of the circular call that @aokolnychyi pointed out, we can either provide a default for `build` or a default for `buildForBatch` / `buildForStreaming`. Anton's approach throws an exception in the old methods, while your approach throws an exception in the new methods.

I think Anton's approach is the right one because it provides backward-compatibility for sources that currently implement the old methods (`buildForBatch`). Those sources are still compatible without being modified. If we removed the implementation of `toBatch` then we would not be able to call `build` unless the source supported it.

Compatibility with existing code is the more important concern.

Reply:
You are right. I overlooked the compatibility issue. Sorry for the wrong suggestion.